### PR TITLE
[Merged by Bors] - Remove redundant table and sparse set component IDs from Archetype

### PR DESCRIPTION
**Pull request by james7132**
# Objective
Archetype is a deceptively large type in memory. It stores metadata about which components are in which storage in multiple locations, which is only used when creating new Archetypes while moving entities.

## Solution
Remove the redundant `Box<[ComponentId]>`s and iterate over the sparse set of component metadata instead. Reduces Archetype's size by 4 usizes (32 bytes on 64-bit systems), as well as the additional allocations for holding these slices.

It'd seem like there's a downside that the origin archetype has it's component metadata iterated over twice when creating a new archetype, but this change also removes the extra `Vec<ArchetypeComponentId>` allocations when creating a new archetype which may amortize out to a net gain here. This change likely negatively impacts creating new archetypes with a large number of components, but that's a cost mitigated by the fact that these archetypal relationships are cached in Edges and is incurred only once for each edge created.

## Additional Context
There are several other in-flight PRs that shrink Archetype:

 - #4800 merges the entities and rows Vecs together (shaves off 24 bytes per archetype) 
 - #4809 removes unique_components and moves it to it's own dedicated storage (shaves off 72 bytes per archetype)

---

## Changelog
Changed: `Archetype::table_components` and `Archetype::sparse_set_components` return iterators instead of slices. `Archetype::new` requires iterators instead of parallel slices/vecs.

## Migration Guide
Do I still need to do this? I really hope people were not relying on the public facing APIs changed here.

### CODE REVIEW

**Comment by alice-i-cecile** [comment]
I'm fine without a migration guide here. Anyone attempting to rely on this is experienced enough to piece it together.

File: crates/bevy_ecs/src/archetype.rs [hunk 1]
```diff
@@ -130,31 +128,25 @@ impl Archetype {
...
+        table_components: impl Iterator<Item = (ComponentId, ArchetypeComponentId)>,
+        sparse_set_components: impl Iterator<Item = (ComponentId, ArchetypeComponentId)>,
     ) -> Self {
-        let mut components =
-            SparseSet::with_capacity(table_components.len() + sparse_set_components.len());
-        for (component_id, archetype_component_id) in
-            table_components.iter().zip(table_archetype_components)
-        {
+        let (min_table, _) = table_components.size_hint();
+        let (min_sparse, _) = table_components.size_hint();
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/archetype.rs:135]:
```suggestion
        let (min_sparse, _) = sparse_set_components.size_hint();
```

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
There's a critical little typo in your code. Once that's fixed this LGTM.

**Review by alice-i-cecile — APPROVED** [review]

**Review by Diddykonga — APPROVED** [review]
Besides for the suggested change from @alice-i-cecile LGTM
A good simplification for sure :)

File: crates/bevy_ecs/src/archetype.rs [hunk 2]
```diff
@@ -453,38 +449,33 @@ impl Archetypes {
...
         let archetypes = &mut self.archetypes;
         let archetype_component_count = &mut self.archetype_component_count;
-        let mut next_archetype_component_id = move || {
-            let id = ArchetypeComponentId(*archetype_component_count);
-            *archetype_component_count += 1;
-            id
-        };
         *self
             .archetype_ids
             .entry(archetype_identity)
```
Comment (cart) [inline crates/bevy_ecs/src/archetype.rs:461]: Not something that needs to be done here, but I think we can completely eliminate the redundant clones / allocations here related to ArchetypeIdentity:

1. Use Hashbrown's raw_entry API to remove the need for an "owning" ArchetypeIdentity when the archetype already exists
2. When the archetype does not exist, create the Archetype first using a non-owning iterator over sparse_set_components and table_components, then use the same allocated lists to construct the final owning ArchetypeIdentity.

**Comment by cart** [comment]
bors retry

**Pull request closed without merging**